src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp

```diff
@@ -0,0 +1,1277 @@
+                                          phaseVolFrac );
+    } );
+  }
+}
+
+
+void ImmiscibleMultiphaseFlow::updateFluidState( ElementSubRegionBase & subRegion ) const
+{
+  GEOS_MARK_FUNCTION;
+
+  updateFluidModel( subRegion );
+  updatePhaseVolumeFraction( subRegion );
+  updatePhaseMass( subRegion );
+  updateRelPermModel( subRegion );
+  updatePhaseMobility( subRegion );
+  updateCapPressureModel( subRegion );
+}
+
+
+void ImmiscibleMultiphaseFlow::updatePhaseMass( ElementSubRegionBase & subRegion ) const
+{
+  GEOS_MARK_FUNCTION;
+
+  string const & solidName = subRegion.getReference< string >( viewKeyStruct::solidNamesString() );
+  string const & fluidName = subRegion.getReference< string >( viewKeyStruct::fluidNamesString() );
+
+  TwoPhaseFluid const & fluid = getConstitutiveModel< TwoPhaseFluid >( subRegion, fluidName );
+  CoupledSolidBase const & solid = getConstitutiveModel< CoupledSolidBase >( subRegion, solidName );
+
+  arrayView1d< real64 const > const volume = subRegion.getElementVolume();
+  arrayView2d< real64 const > const porosity = solid.getPorosity();
+
+  arrayView2d< real64 const, immiscibleFlow::USD_PHASE > const phaseVolFrac= subRegion.getField< fields::immiscibleMultiphaseFlow::phaseVolumeFraction >();
+  arrayView3d< real64 const, multifluid::USD_PHASE > phaseDens = fluid.phaseDensity();
+  arrayView2d< real64, immiscibleFlow::USD_PHASE > const phaseMass = subRegion.getField< fields::immiscibleMultiphaseFlow::phaseMass >();
+
+  // Might be needed for geomechanics????? if so, need to change the accumulation as well?
+  //arrayView1d< real64 > const deltaVolume = subRegion.getField< fields::flow::deltaVolume >();
+
+  forAll< parallelDevicePolicy<> >( subRegion.size(), [=] GEOS_HOST_DEVICE ( localIndex const ei )
+  {
+    real64 const poreVolume = volume[ei] * porosity[ei][0];
+    for( integer ip = 0; ip < 2; ++ip )
+    {
+      phaseMass[ei][ip] = poreVolume * phaseVolFrac[ei][ip] * phaseDens[ei][0][ip];
+    }
+  } );
+}
+
+
+void ImmiscibleMultiphaseFlow::updatePhaseMobility( ObjectManagerBase & dataGroup ) const
+{
+  GEOS_MARK_FUNCTION;
+
+  // note that the phase mobility computed here also includes phase density
+  string const & fluidName = dataGroup.getReference< string >( viewKeyStruct::fluidNamesString() );
+  TwoPhaseFluid const & fluid = getConstitutiveModel< TwoPhaseFluid >( dataGroup, fluidName );
+
+  string const & relpermName = dataGroup.getReference< string >( viewKeyStruct::relPermNamesString() );
+  RelativePermeabilityBase const & relperm = getConstitutiveModel< RelativePermeabilityBase >( dataGroup, relpermName );
+
+  immiscibleMultiphaseKernels::
+    PhaseMobilityKernelFactory::
+    createAndLaunch< parallelDevicePolicy<> >( m_numPhases,
+                                               dataGroup,
+                                               fluid,
+                                               relperm );
+}
+
+void ImmiscibleMultiphaseFlow::initializeFluidState( MeshLevel & mesh,
```

Comment:
check https://github.com/GEOS-DEV/GEOS/pull/3393, i hope it is not going to be hard to merge